### ISSUE TITLE
Descriptions for nested schemas are missing

### ISSUE THREAD

**Issue by Integralist** [hashicorp/terraform-plugin-docs#28]
Here's an example schema definition with a nested schema under the 'rules' field:

```hcl
		Schema: map[string]*schema.Schema{
			"publishers": {
				Type:        schema.TypeList,
				Optional:    true,
				Description: "A list of publishers to be used as filters for the data set.",
				Elem:        &schema.Schema{Type: schema.TypeString},
			},
			"tags": {
				Type:        schema.TypeList,
				Optional:    true,
				Description: "A list of tags to be used as filters for the data set.",
				Elem:        &schema.Schema{Type: schema.TypeString},
			},
			"exclude_modsec_rule_ids": {
				Type:        schema.TypeList,
				Optional:    true,
				Description: "A list of modsecurity rules IDs to be excluded from the data set.",
				Elem:        &schema.Schema{Type: schema.TypeInt},
			},
			"rules": {
				Type:        schema.TypeList,
				Computed:    true,
				Description: "The list of rules that results from any given combination of filters.",
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"modsec_rule_id": {
							Type:        schema.TypeInt,
							Required:    true,
							Description: "The modsecurity rule ID.",
						},
						"latest_revision_number": {
							Type:        schema.TypeInt,
							Required:    true,
							Description: "The modsecurity rule's latest revision.",
						},
						"type": {
							Type:        schema.TypeString,
							Computed:    true,
							Description: "The modsecurity rule's type.",
						},
					},
				},
			},
		},

```

In the generated schema output (see below example, from running `tfplugindocs generate`) we can see the `Description` fields from the top-level `Schema` types are included in the output, but the nested Schema Description fields are not included (look `<MISSING DESCRIPTION>`)...

```
<!-- schema generated by tfplugindocs -->
## Schema

### Optional

- **exclude_modsec_rule_ids** (List of Number) A list of modsecurity rules IDs to be excluded from the data set.
- **id** (String) The ID of this resource.
- **publishers** (List of String) A list of publishers to be used as filters for the data set.
- **tags** (List of String) A list of tags to be used as filters for the data set.

### Read-only

- **rules** (List of Object) The list of rules that results from any given combination of filters. (see [below for nested schema](#nestedatt--rules))

<a id="nestedatt--rules"></a>
### Nested Schema for `rules`

Read-only:

- **latest_revision_number** (Number) <MISSING DESCRIPTION>
- **modsec_rule_id** (Number) <MISSING DESCRIPTION>
- **type** (String) <MISSING DESCRIPTION>
```

**Comment by Integralist** [comment]
Interestingly this isn't consistent behaviour 🤔 

The following nested schema data _was_ getting the descriptions templated into the Markdown as we would have hoped...

```
Schema: map[string]*schema.Schema{
	"service_id": {
		Type:        schema.TypeString,
		Required:    true,
		ForceNew:    true,
		Description: "Service Id",
	},

	"acl_id": {
		Type:        schema.TypeString,
		Required:    true,
		ForceNew:    true,
		Description: "ACL Id",
	},
	"entry": {
		Type:        schema.TypeSet,
		Optional:    true,
		Description: "ACL Entries",
		MaxItems:    gofastly.MaximumACLSize,
		Elem: &schema.Resource{
			Schema: map[string]*schema.Schema{
				"id": {
					Type:        schema.TypeString,
					Description: "",
					Computed:    true,
				},
				"ip": {
					Type:        schema.TypeString,
					Description: "An IP address that is the focus for the ACL",
					Required:    true,
				},
				"subnet": {
					Type:        schema.TypeString,
					Optional:    true,
					Description: "An optional subnet mask applied to the IP address",
				},
				"negated": {
					Type:        schema.TypeBool,
					Optional:    true,
					Default:     false,
					Description: "A boolean that will negate the match if true",
				},
				"comment": {
					Type:        schema.TypeString,
					Optional:    true,
					Description: "A personal freeform descriptive note",
				},
			},
		},
	},
},
```

The generated Markdown shows the Description field values are being used:

```
<a id="nestedblock--entry"></a>
### Nested Schema for `entry`

Required:

- **ip** (String) An IP address that is the focus for the ACL

Optional:

- **comment** (String) A personal freeform descriptive note
- **negated** (Boolean) A boolean that will negate the match if true
- **subnet** (String) An optional subnet mask applied to the IP address

Read-only:

- **id** (String) The ID of this resource.
```

Also interesting is that it has been able to generate an adequate description for the 'read-only' ID field (which in the Schema was left with a blank string for its value) 🤔

**Comment by paultyng** [comment]
The explanation for this goes in to Terraform internals. And just an aside, to debug you may want to start with running `terraform providers schema -json`, this is the JSON file the tool works with so easier to debug in there.

The key there is that this is a `List of Object`, not a `Block List`. Those are two distinct different types in HCL/Terraform. A block has attributes, but an object type has fields (in syntax blocks look like `thing {` but an object attribute looks like `thing = {`). 

Object fields don't have metadata currently, so they can't contain attributes like description/sensitive, etc. `helper/schema` was written pretty early in the evolution of Terraform and its abstraction doesn't fully match the concepts of schema [going across the wire](https://github.com/hashicorp/terraform/blob/master/docs/plugin-protocol/tfplugin5.2.proto). 

We do have some thoughts on how to address this for document generation, basically decoupling the the generation from the schema JSON so that it can potentially have richer information supplied, but its probably a ways down the road at this point.

**Comment by Integralist** [comment]
Thanks @paultyng for the feedback/update 👍🏻

**Comment by paultyng** [comment]
Also worth noting there is an ongoing internal project to look at normalizing these things so you don't have the trade-offs, if/when that lands we can update the SDK and generator to use it, but probably a ways out.

Unfortunately for the time being this probably means you will have a few docs you'll need to write schema for by hand (or just `sed` in some descriptions or something).

**Comment by elXaptation** [comment]
@paultyng Thank you for the details. Is this work/effort being tracked anywhere I can follow along?

**Comment by pontovinte** [comment]
I was about to ask the same, is there someone actively working on this?

**Comment by Macaber** [comment]
I was about to ask the same, is there someone actively working on this?

**Comment by oscarhermoso** [comment]
I _think_ I have discovered one way to reproduce this bug... it occurs when prefixing the schema attribute with the `&`  address-of operator.

TL;DR; if you do this, it fixes the issue:

```diff
- s.Attributes["forwarding_rules"] = &schema.ListNestedAttribute{
+ s.Attributes["forwarding_rules"] = schema.ListNestedAttribute{
```

# Steps to reproduce

I have a `LoadBalancerResourceSchema` which was code-gen'd by [terraform-plugin-codegen-openapi](https://github.com/hashicorp/terraform-plugin-codegen-openapi).

I want to override it slightly, copying all properties from the existing schema except I've added a `Default` to the `ListNestedAttribute`.

```hcl
func loadBalancerSchema(ctx context.Context) schema.Schema {
	s := resources.LoadBalancerResourceSchema(ctx)
	
	# ...

	// Overrides
	forwardingRules := &s.Attributes["forwarding_rules"].(schema.ListNestedAttribute)
	s.Attributes["forwarding_rules"] = &schema.ListNestedAttribute{
		NestedObject:        forwardingRules.NestedObject,
		Description:         forwardingRules.GetDescription(),
		MarkdownDescription: forwardingRules.GetMarkdownDescription(),
		Required:            false,
		Optional:            true,
		Computed:            true,
		Default: listdefault.StaticValue(types.ListValueMust(
			types.ObjectType{
				AttrTypes: resources.ForwardingRulesValue{}.AttributeTypes(ctx),
			},
			[]attr.Value{
				types.ObjectValueMust(
					map[string]attr.Type{
						"entry_protocol": types.StringType,
					},
					map[string]attr.Value{
						"entry_protocol": types.StringValue("http"),
					},
				),
			},
		)),
	}
```

# Expected tfplugindocs output

```md
<!-- schema generated by tfplugindocs -->
## Schema

### Required

- `id` (Number) The ID of the load balancer to fetch.

### Read-Only

- `forwarding_rules` (Attributes List) The rules that control which traffic the load balancer will forward to servers in the pool. Leave null to accept a default "HTTP" only forwarding rule. (see [below for nested schema](#nestedatt--forwarding_rules))
- `health_check` (Object) The rules that determine which servers are considered 'healthy' and in the server pool for the load balancer. Leave this null to accept appropriate defaults based on the forwarding_rules. (see [below for nested schema](#nestedatt--health_check))
- `ip` (String) The IPv4 address of the load balancer.
- `name` (String) The hostname of the load balancer.
- `region` (String) Leave null to create an anycast load balancer.
- `server_ids` (List of Number) A list of server IDs to assign to this load balancer.

<a id="nestedatt--forwarding_rules"></a>
### Nested Schema for `forwarding_rules`

Read-Only:

- `entry_protocol` (String) The protocol that traffic must match for the load balancer to forward it. Valid values are "http" and "https".
```

# Actual tfplugindocs output

Notice the the fact that `forwarding_rules` is now a (List of Object), and the lack of description on the `entry_protocol` in the Nested Schema.

```md
<!-- schema generated by tfplugindocs -->
## Schema

### Required

- `id` (Number) The ID of the load balancer to fetch.

### Read-Only

- `forwarding_rules` (List of Object) The rules that control which traffic the load balancer will forward to servers in the pool. Leave null to accept a default "HTTP" only forwarding rule. (see [below for nested schema](#nestedatt--forwarding_rules))
- `health_check` (Object) The rules that determine which servers are considered 'healthy' and in the server pool for the load balancer. Leave this null to accept appropriate defaults based on the forwarding_rules. (see [below for nested schema](#nestedatt--health_check))
- `ip` (String) The IPv4 address of the load balancer.
- `name` (String) The hostname of the load balancer.
- `region` (String) Leave null to create an anycast load balancer.
- `server_ids` (List of Number) A list of server IDs to assign to this load balancer.

<a id="nestedatt--forwarding_rules"></a>
### Nested Schema for `forwarding_rules`

Read-Only:

- `entry_protocol` (String)
```

# The fix?

After removing the `&` - I started generating the expected output. 🎉

```diff
	// Overrides
-	forwardingRules := &s.Attributes["forwarding_rules"].(schema.ListNestedAttribute)
+	forwardingRules := s.Attributes["forwarding_rules"].(schema.ListNestedAttribute)
	s.Attributes["forwarding_rules"] = &schema.ListNestedAttribute{
```

So why the question mark in the heading?

Well... I'm using `terraform-plugin-framework`, whereas a lot of the linked threads here are using `terraform-plugin-sdk/v2` - so this fix won't actually apply to most of the repos linked to this thread.

Furthermore, this might not even be a bug with `terraform-plugin-docs` at all, it might be a bug with a util function that I'm using to generate a data source schema from a resource schema. https://github.com/oscarhermoso/terraform-provider-binarylane/blob/424828123df07fcb41f186f089b491c47c8b3508/internal/provider/util.go#L20

Either way, the outcome was so simiar that I felt it was worth sharing my findings on this thread to hopefully help someone else who stumbles upon this issue & finds it at the top of their Google search like me.

**Comment by deiga** [comment]
Further data point for investigation:

<details>
<summary>Template</summary>

```markdown
---
page_title: "{{.Name}} ({{.Type}}) - {{.RenderedProviderName}}"
subcategory: ""
description: |-
{{ .Description | plainmarkdown | trimspace | prefixlines "  " }}
---

# {{.Name}} ({{.Type}})

{{ .Description | trimspace }}

## Notes

- This resource is **organization-only** and cannot be used with individual accounts.
- Image IDs for GitHub-owned images are numeric strings (e.g., "2306" for Ubuntu Latest 24.04), not names like "ubuntu-latest".
- Deletion of hosted runners is asynchronous. The provider will poll for up to 10 minutes (configurable via timeouts) to confirm deletion.
- Runner creation and updates may take several minutes as GitHub provisions the infrastructure.

## Getting Available Images and Sizes
...

{{ if .HasExamples -}}
## Example Usage

{{- range .ExampleFiles }}

{{- $parts := split . "resource_" -}}
{{- $lastItem := index $parts 1 }}
{{- $fileNameParts := split $lastItem "." }}
{{- $fileName := index $fileNameParts 0 }}

### {{ $fileName | title }} Usage

{{ tffile . }}
{{- end }}
{{- end }}

{{ .SchemaMarkdown | trimspace }}
{{- if or .HasImport .HasImportIDConfig .HasImportIdentityConfig }}

## Import

Import is supported using the following syntax:
{{- end }}
{{- if .HasImportIdentityConfig }}

In Terraform v1.12.0 and later, the [`import` block](https://developer.hashicorp.com/terraform/language/import) can be used with the `identity` attribute, for example:

{{tffile .ImportIdentityConfigFile }}

{{ .IdentitySchemaMarkdown | trimspace }}
{{- end }}
{{- if .HasImportIDConfig }}

In Terraform v1.5.0 and later, the [`import` block](https://developer.hashicorp.com/terraform/language/import) can be used with the `id` attribute, for example:

{{tffile .ImportIDConfigFile }}
{{- end }}
{{- if .HasImport }}

The [`terraform import` command](https://developer.hashicorp.com/terraform/cli/commands/import) can be used, for example:

{{codefile "shell" .ImportFile }}
{{- end }}

```
</details>

<details>
<summary>Schema</summary>

```
Description: "This resource allows you to create and manage GitHub-hosted runners within your GitHub organization. You must have admin access to an organization to use this resource.",
		Schema: map[string]*schema.Schema{
			"name": {
				Type:     schema.TypeString,
				Required: true,
				ValidateDiagFunc: validation.ToDiagFunc(validation.All(
					validation.StringLenBetween(1, 64),
					validation.StringMatch(
						regexp.MustCompile(`^[a-zA-Z0-9._-]+$`),
						"name may only contain alphanumeric characters, '.', '-', and '_'",
					),
				)),
				Description: "Name of the hosted runner. Must be between 1 and 64 characters and may only contain upper and lowercase letters a-z, numbers 0-9, '.', '-', and '_'.",
			},
			"image": {
				Type:        schema.TypeList,
				Required:    true,
				ForceNew:    true,
				MaxItems:    1,
				Description: "Image configuration for the hosted runner. Cannot be changed after creation.",
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"id": {
							Type:        schema.TypeString,
							Required:    true,
							Description: "The image ID.",
						},
						"source": {
							Type:             schema.TypeString,
							Optional:         true,
							Default:          "github",
							ValidateDiagFunc: validation.ToDiagFunc(validation.StringInSlice([]string{"github", "partner", "custom"}, false)),
							Description:      "The image source (github, partner, or custom).",
						},
						"size_gb": {
							Type:        schema.TypeInt,
							Computed:    true,
							Description: "The size of the image in GB.",
						},
					},
				},
			},
			"size": {
				Type:        schema.TypeString,
				Required:    true,
				Description: "Machine size for the hosted runner (e.g., `4-core`, `8-core`). Can be updated to scale the runner. To list available sizes, use the GitHub API: `GET /orgs/{org}/actions/hosted-runners/machine-sizes`.",
			},
			"runner_group_id": {
				Type:        schema.TypeInt,
				Required:    true,
				Description: "The ID of the runner group to assign this runner to.",
			},
			"maximum_runners": {
				Type:             schema.TypeInt,
				Optional:         true,
				Computed:         true,
				ValidateDiagFunc: validation.ToDiagFunc(validation.IntAtLeast(1)),
				Description:      "Maximum number of runners to scale up to. Runners will not auto-scale above this number. Use this setting to limit costs.",
			},
			"public_ip_enabled": {
				Type:        schema.TypeBool,
				Optional:    true,
				Default:     false,
				Description: "Whether to enable static public IP for the runner. Note there are account limits. To list limits, use the GitHub API: `GET /orgs/{org}/actions/hosted-runners/limits`. Defaults to false.",
			},
			"image_version": {
				Type:        schema.TypeString,
				Optional:    true,
				Description: "The version of the runner image to deploy. This is only relevant for runners using custom images.",
			},
			"image_gen": {
				Type:        schema.TypeBool,
				Optional:    true,
				ForceNew:    true,
				Default:     false,
				Description: "Whether this runner should be used to generate custom images. Cannot be changed after creation.",
			},
			"id": {
				Type:        schema.TypeString,
				Computed:    true,
				Description: "The hosted runner ID.",
			},
			"status": {
				Type:        schema.TypeString,
				Computed:    true,
				Description: "Current status of the runner.",
			},
			"machine_size_details": {
				Type:        schema.TypeList,
				Computed:    true,
				Description: "Detailed machine size specifications.",
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"id": {
							Type:        schema.TypeString,
							Computed:    true,
							Description: "Machine size ID.",
						},
						"cpu_cores": {
							Type:        schema.TypeInt,
							Computed:    true,
							Description: "Number of CPU cores.",
						},
						"memory_gb": {
							Type:        schema.TypeInt,
							Computed:    true,
							Description: "Memory in GB.",
						},
						"storage_gb": {
							Type:        schema.TypeInt,
							Computed:    true,
							Description: "Storage in GB.",
						},
					},
				},
			},
			"platform": {
				Type:        schema.TypeString,
				Computed:    true,
				Description: "Platform of the runner.",
			},
			"public_ips": {
				Type:        schema.TypeList,
				Computed:    true,
				Description: "List of public IP ranges assigned to this runner.",
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"enabled": {
							Type:        schema.TypeBool,
							Computed:    true,
							Description: "Whether this IP range is enabled.",
						},
						"prefix": {
							Type:        schema.TypeString,
							Computed:    true,
							Description: "IP address prefix.",
						},
						"length": {
							Type:        schema.TypeInt,
							Computed:    true,
							Description: "Subnet length.",
						},
					},
				},
			},
			"last_active_on": {
				Type:        schema.TypeString,
				Computed:    true,
				Description: "Timestamp when the runner was last active.",
			},
		},
```
</details>

<details>
<summary>Generated doc</summary>

```markdown
---
page_title: "github_actions_hosted_runner (Resource) - GitHub"
subcategory: ""
description: |-
  This resource allows you to create and manage GitHub-hosted runners within your GitHub organization. You must have admin access to an organization to use this resource.
---

# github_actions_hosted_runner (Resource)

This resource allows you to create and manage GitHub-hosted runners within your GitHub organization. You must have admin access to an organization to use this resource.

## Notes

- This resource is **organization-only** and cannot be used with individual accounts.
- Image IDs for GitHub-owned images are numeric strings (e.g., "2306" for Ubuntu Latest 24.04), not names like "ubuntu-latest".
- Deletion of hosted runners is asynchronous. The provider will poll for up to 10 minutes (configurable via timeouts) to confirm deletion.
- Runner creation and updates may take several minutes as GitHub provisions the infrastructure.

## Getting Available Images and Sizes

To get a list of available images:

```bash
curl -H "Authorization: Bearer YOUR_TOKEN" \
     -H "Accept: application/vnd.github+json" \
     https://api.github.com/orgs/YOUR_ORG/actions/hosted-runners/images/github-owned
```

To get available machine sizes:

```bash
curl -H "Authorization: Bearer YOUR_TOKEN" \
     -H "Accept: application/vnd.github+json" \
     https://api.github.com/orgs/YOUR_ORG/actions/hosted-runners/machine-sizes
```

## Example Usage

### Advanced Usage

```terraform
resource "github_actions_runner_group" "example" {
  name       = "example-runner-group"
  visibility = "all"
}

resource "github_actions_hosted_runner" "example" {
  name = "example-hosted-runner"

  image {
    id     = "2306"
    source = "github"
  }

  size              = "8-core"
  runner_group_id   = github_actions_runner_group.example.id
  maximum_runners   = 10
  public_ip_enabled = true
}
```


## Schema

### Required

- `image` (Block List, Min: 1, Max: 1) Image configuration for the hosted runner. Cannot be changed after creation. (see [below for nested schema](#nestedblock--image))
- `name` (String) Name of the hosted runner. Must be between 1 and 64 characters and may only contain upper and lowercase letters a-z, numbers 0-9, '.', '-', and '_'.
- `runner_group_id` (Number) The ID of the runner group to assign this runner to.
- `size` (String) Machine size for the hosted runner (e.g., `4-core`, `8-core`). Can be updated to scale the runner. To list available sizes, use the GitHub API: `GET /orgs/{org}/actions/hosted-runners/machine-sizes`.

### Optional

- `image_gen` (Boolean) Whether this runner should be used to generate custom images. Cannot be changed after creation.
- `image_version` (String) The version of the runner image to deploy. This is only relevant for runners using custom images.
- `maximum_runners` (Number) Maximum number of runners to scale up to. Runners will not auto-scale above this number. Use this setting to limit costs.
- `public_ip_enabled` (Boolean) Whether to enable static public IP for the runner. Note there are account limits. To list limits, use the GitHub API: `GET /orgs/{org}/actions/hosted-runners/limits`. Defaults to false.
- `timeouts` (Block, Optional) (see [below for nested schema](#nestedblock--timeouts))

### Read-Only

- `id` (String) The hosted runner ID.
- `last_active_on` (String) Timestamp when the runner was last active.
- `machine_size_details` (List of Object) Detailed machine size specifications. (see [below for nested schema](#nestedatt--machine_size_details))
- `platform` (String) Platform of the runner.
- `public_ips` (List of Object) List of public IP ranges assigned to this runner. (see [below for nested schema](#nestedatt--public_ips))
- `status` (String) Current status of the runner.

<a id="nestedblock--image"></a>
### Nested Schema for `image`

Required:

- `id` (String) The image ID.

Optional:

- `source` (String) The image source (github, partner, or custom).

Read-Only:

- `size_gb` (Number) The size of the image in GB.


<a id="nestedblock--timeouts"></a>
### Nested Schema for `timeouts`

Optional:

- `delete` (String)


<a id="nestedatt--machine_size_details"></a>
### Nested Schema for `machine_size_details`

Read-Only:

- `cpu_cores` (Number)
- `id` (String)
- `memory_gb` (Number)
- `storage_gb` (Number)


<a id="nestedatt--public_ips"></a>
### Nested Schema for `public_ips`

Read-Only:

- `enabled` (Boolean)
- `length` (Number)
- `prefix` (String)

```

**Comment by stevehipwell** [comment]
As far as computed blocks using `terraform-plugin-sdk/v2`; this currently looks impossible to resolve given the fact that the descriptions are being dropped due to the schema defining the types as objects. I suspect that switching to the `terraform-plugin-framework` would fix this due to the way it supports mapping the schema.

That said, I can't see why if the schema is unfixable that this tool couldn't augment the schema metadata when operating on the raw code?

**Comment by magodo** [comment]
For people struggle in this issue (and many others) like me, I've created a library to solve all of them in https://github.com/magodo/terraform-plugin-framework-docs (though only works for `terraform-plugin-framework`).

For this specific one, the user can add custom descriptions for the attributes in `ObjectAttribute` like: https://github.com/magodo/terraform-plugin-framework-docs/blob/f8462799882f64a83f1e2410bc71be229dec98fe/internal/testprovider/resource.go#L148 which then will be rendered into: https://github.com/magodo/terraform-plugin-framework-docs/blob/f8462799882f64a83f1e2410bc71be229dec98fe/testdata/resource_complete.md#nested-schema-for-objectfoo

**Comment by stevehipwell** [comment]
For anyone stuck on `terraform-plugin-sdk/v2` the following snippet (to replace `{{ .SchemaMarkdown | trimspace }}`) will allow you to generate the schema without descriptions in a comment block so you can at least track the structural changes and keep them documented.

```tmpl
<!--
{{- $schema := .SchemaMarkdown | trimspace -}}
{{- range $line := split $schema "\n" }}
{{- if ne (trimspace $line) "<!-- schema generated by tfplugindocs -->" }}
{{ $line }}
{{- end }}
{{- end }}
-->
```
